File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        long long prev = LLONG_MIN;
        stack<TreeNode*> s;
        TreeNode* curr = root;

        while (!s.empty() || curr) {
            while (curr) {
                s.push(curr);
                curr = curr->left;
            }

            curr = s.top();
            s.pop();
            if (curr->val <= prev) {
                return false;
            }

            prev = curr->val;
            curr = curr->right;
                
        }

        return true;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root: TreeNode | null, k: number): number {
        // in order BST always produces increasing list
        let count = 0;
        const stack: TreeNode[] = [];
        let curr: TreeNode | null = root;

        while (stack.length > 0 || curr !== null) {
            // left
            while (curr !== null) {
                stack.push(curr);
                curr = curr.left;
            }

            // parent
            curr = stack.pop()!;
            if (count == k-1) {
                return curr.val;
            }
            count++;

            // right
            curr = curr.right;
        }

        return curr.val;
    }
}
